packages/alfa-frontier/src/frontier.ts

```diff
@@ -0,0 +1,370 @@
+import { Equatable } from "@siteimprove/alfa-equatable";
+import { Serializable } from "@siteimprove/alfa-json";
+import { Option, None } from "@siteimprove/alfa-option";
+import { Predicate } from "@siteimprove/alfa-predicate";
+
+import * as json from "@siteimprove/alfa-json";
+
+const { equals, property } = Predicate;
+
+export class Frontier implements Equatable, Serializable {
+  public static of(
+    scope: string | URL,
```

Comment:
Actually, `Frontier` and `Item` are pretty close to being a generic queue management system, not just for URLs.
Need to add a type parameter to both, and change `scope` to a `T => bool`. The rest of `Frontier` is pretty much agnostic to the actual stuff in it, and it looks like `Item` only need to know that it's URLs for its `matches` method which could as well be avoided by requiring `T` to be `Equatable`.

Reply:
It's really not a generic queue of any sorts though as every operation has semantics specific to a crawl frontier. Sure, you might be able to abstract over the type of things stored in the frontier, though I don't see any use for it as its only purpose is to act as a URL frontier. Any attempt to abstract it beyond that seem like needless complexity 🤔

Reply:
It's also work keeping in mind that this is a structure that callers will interact with directly and so any additional layers of indirection we add will be layers that callers will have to unwind.